integration/tests/cook/test_multi_user.py

```diff
@@ -113,6 +113,63 @@ def test_multi_user_usage(self):
                 with self.user_factory.admin():
                     util.kill_jobs(self.cook_url, all_job_uuids, assert_response=False)
 
+    @pytest.mark.serial
+    def test_user_pool_rate_limit(self):
+        settings_dict = util.settings(self.cook_url)
+        if settings_dict.get('rate-limit', {}).get('per-user-per-pool-job-launch', {}) is None:
+            pytest.skip("Can't test job launch rate limit without launch rate limit set.")
+        if settings_dict.get('rate-limit', {}).get('per-user-per-pool-job-launch', {}).get('enforce?', None) is not True:
+            pytest.skip("Can't test job launch rate limit without launch rate limit set to enforcing.")
+
+        bucket_size = 5
+        token_rate = 1
+
+        admin = self.user_factory.admin()
+        user = self.user_factory.new_user()
+        job_uuids = []
+
+        try:
+            with admin:
+                resp = util.set_limit(self.cook_url, 'quota', user.name, bucket_size=bucket_size, token_rate=token_rate)
+                self.assertEqual(resp.status_code, 201, resp.text)
+
+            with user:
+                jobspec = {"command": "sleep 240", 'cpus': 0.03, 'mem': 32}
+
+                self.logger.info(f'Submitting initial batch of {bucket_size - 1} jobs')
+                initial_uuids, initial_response = util.submit_jobs(self.cook_url, jobspec, bucket_size - 1)
+                job_uuids.extend(initial_uuids)
+                self.assertEqual(201, initial_response.status_code, msg=initial_response.content)
+
+                util.wait_for_jobs(self.cook_url, job_uuids, 'running')
+
+                self.logger.info(f'Submitting subsequent batch of {bucket_size - 1} jobs')
+                subsequent_uuids, subsequent_response = util.submit_jobs(self.cook_url, jobspec, bucket_size - 1)
+                job_uuids.extend(subsequent_uuids)
+                self.assertEqual(201, subsequent_response.status_code, msg=subsequent_response.content)
+
+                def submit_jobs():
+                    pass
+
+                def is_rate_limit_triggered(_):
+                    jobs1 = util.query_jobs(self.cook_url, True, uuid=job_uuids).json()
+                    running_jobs = [j for j in jobs1 if j['status'] == 'running']
+                    waiting_jobs = [j for j in jobs1 if j['status'] == 'waiting']
+                    self.logger.debug(f'There are {len(waiting_jobs)} waiting jobs')
+                    return len(waiting_jobs) > 0 and len(running_jobs) >= bucket_size
+
+                util.wait_until(submit_jobs, is_rate_limit_triggered, 120000, 5000)
+                jobs2 = util.query_jobs(self.cook_url, True, uuid=job_uuids).json()
+                running_jobs = [j for j in jobs2 if j['status'] == 'running']
+                self.assertGreaterEqual(len(running_jobs), bucket_size)
+                self.assertLessEqual(len(running_jobs), bucket_size + 5)
```

Comment:
We may launch a (small) number of extra jobs because the token buffer may give a few tokens during the waiting period. I need to account for those. But, I think that +5  is wrong now that I can set the quota (and am not stuck with a predetermined global range). It should be 2x bucket size.